gateway/src/apicast/cli/environment.lua

```diff
@@ -51,28 +51,52 @@ local function detect_kubernetes()
   return secrets or resty_env.value('KUBERNETES_PORT')
 end
 
+-- CPU shares in Cgroups v1 or converted from weight in Cgroups v2 in millicores
 local function cpu_shares()
-  if not detect_kubernetes() then return end
-
   local shares
-  local file = open('/sys/fs/cgroup/cpu/cpu.shares')
 
-  if file then
-    shares = file:read('*n')
+  -- This check is from https://github.com/kubernetes/kubernetes/blob/release-1.27/test/e2e/node/pod_resize.go#L305-L314
+  -- alternatively, this method can be used: https://kubernetes.io/docs/concepts/architecture/cgroups/#check-cgroup-version
+  -- (`stat -fc %T /sys/fs/cgroup/` returns `cgroup2fs` or `tmpfs`)
+  if pl_path.exists("/sys/fs/cgroup/cgroup.controllers") then
+    -- Cgroups v2
+    ngx.log(ngx.DEBUG, "detecting cpus in Cgroups v2")
+    -- Using the formula from https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/kubelet/cm/cgroup_manager_linux.go#L570-L574
+    local file = open('/sys/fs/cgroup/cpu.weight')
+
+    if file then
+      local weight = file:read('*n')
+      shares = (((weight - 1) * 262142) / 9999) + 2
+
+      file:close()
```

Comment:
(just an arguable nitpick)
Wondering if it makes sense to `close` right after reading from the file.
There seem to be no "unsafe" operations on that line that could cause an error, but I think the sooner we close, the better.